sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c

```diff
@@ -6151,10 +6165,10 @@ dtrace_dif_emulate(dtrace_difo_t *difo, dtrace_mstate_t *mstate,
 			regs[rd] = regs[r1] - regs[r2];
 			break;
 		case DIF_OP_ADD:
-			regs[rd] = regs[r1] + regs[r2];
+			regs[rd] = (uint64_t)regs[r1] + regs[r2];
```

Comment:
This line breaks the most of D scripts by making it impossible to retrieve a capability from it. The DTrace compiler will emit the following sequence for any kind of pointer following:

```
syscall::read:entry { x = curthread->td_tid; }

OFF OPCODE      INSTRUCTION
00: 29010001    ldgs DT_VAR(256), %r1		! DT_VAR(256) = "curthread"
01: 25000002    setx DT_INTEGER[0], %r2		! 0xa0
02: 07010201    add  %r1, %r2, %r1
03: 1e010001    ldsw [%r1], %r1
04: 23000001    ret  %r1
```

Notice how we have a `ldgs DT_VAR(256), %r1` which loads the value of the `curthread` variable into %r1, making it of type `struct thread *`. Later on, we have `add  %r1, %r2, %r1`, which adds a `uint64_t` to `%r1` in order to reach the `td_tid` struct member, followed by a `ldsw [%r1], %r1`, which dereferences whatever is in `%r1` and stores the result in `%r1`. Since this casts the capability to an `uint64_t`, this can never be dereferenced in an `ldub/lduh/lduw/ldsb/ldsh/ldsw/ldx` instruction, since it is an invalid capability. The same might be true for some other instructions.

Reply:
True, I think this is a left out from a previous change. Now `regs` contains `uintcap_t`, sho the cast should not be needed anymore